### PR TITLE
[FSDP] compile compute and CI with @test_compiled_fsdp

### PR DESCRIPTION
**Pull request by weifengpy**
goal: all unit tests for eager. we want to test torch.compile by default

this PR adds ``@test_compiled_fsdp(compile_compute_on_module=None/TransformerBlock)`` to unit tests. now it's compiling compute-only as follows. 

```
module.compile() # include user registered hooks if any
fully_shard(module)
```

torch.compile does not work following component yet
* compiling AC
* compiling reshard_after_forward=2
* delayed_all_gather, delayed_reduce_scatter

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 1]
```diff
@@ -346,6 +349,7 @@ def test_non_root_forward_backward(self):
         self.assertEqual(ref_nonroot_loss, nonroot_loss)
         self.assertEqual(ref_model(inp).sum(), model(inp).sum())
 
+    @test_compiled_fsdp()
     def test_multi_forward_module(self):
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:353]: I realized I missed a `@skip_if_lt_x_gpu(2)` for this test. Could you sneak it into this PR 😅

File: torch/testing/_internal/common_fsdp.py [hunk 2]
```diff
@@ -1352,6 +1355,41 @@ def _test_fsdp_parity(
...
 
+def test_compiled_fsdp():
+    class FullyShardPatch(Enum):
+        @staticmethod
+        def fully_shard_with_compiled_compute(*args, **kwargs):
+            args[0].forward = torch.compile(args[0].forward)
+            return torch.distributed._composable.fsdp.fully_shard(*args, **kwargs)
+
+        @staticmethod
+        def fully_shard_with_full_compile(*args, **kwargs):
```
Comment (awgu) [inline torch/testing/_internal/common_fsdp.py:1366]: I am not sure if this makes sense since for full compile, we would only apply it to the root module. However, IIUC, this would do like `torch.compile(fully_shard(module))` for each `fully_shard` call, including non-root.

File: test/distributed/_composable/fsdp/test_fully_shard_training.py (hunk 1 above)
Comment (weifengpy) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:353]: will add it to the PR

File: torch/testing/_internal/common_fsdp.py (hunk 2 above)
Comment (weifengpy) [reply torch/testing/_internal/common_fsdp.py:1366]: good catch! will think about a better way but will remove ``fully_shard_with_full_compile`` from this PR

File: torch/testing/_internal/common_fsdp.py [hunk 3]
```diff
@@ -1352,6 +1355,36 @@ def _test_fsdp_parity(
             )
 
 
+def test_compiled_fsdp():
+    class FullyShardPatch(Enum):
+        @staticmethod
+        def fully_shard_with_compiled_compute(*args, **kwargs):
+            args[0].forward = torch.compile(args[0].forward)
```
[the remark this answers is not in the document]
Comment (weifengpy) [reply torch/testing/_internal/common_fsdp.py:1362]: wow! just get to know nn.Module.compile. will give it a try

[the remark this answers is not in the document]
Comment (weifengpy) [reply torch/testing/_internal/common_fsdp.py]: updated the PR with ``nn.Module.compile``

[the remark this answers is not in the document]
Comment (awgu) [reply torch/testing/_internal/common_fsdp.py]: Another thing to be wary of: https://github.com/pytorch/pytorch/issues/117584 

In practice for transformers, we mainly care about compiling the transformer blocks. The embedding and final projection are less important (since they consist of a smaller portion of the FLOPs). We should be wary of a case where, if we compile the root module and the transformer blocks, we error, but we could have just only compiled the transformer blocks.

For example, for the fp8 work right now, this is the case. We cannot compile the root but can compile the transformer blocks.

[the remark this answers is not in the document]
Comment (weifengpy) [reply torch/testing/_internal/common_fsdp.py]: what do you think of ``@test_compiled_fsdp(compile_on_module=TransformerBlock)`` ? whether or not compiling root module sounds like model specific

[the remark this answers is not in the document]
Comment (awgu) [reply torch/testing/_internal/common_fsdp.py]: Specifying which modules to compile seems reasonable to me!

[the remark this answers is not in the document]
Comment (weifengpy) [reply torch/testing/_internal/common_fsdp.py]: updated PR. For transformers, only compile blocks

**Comment by wanchaol** [comment]
This is an interesting PR! We could do sth similar to turn on compile for all DTensor tests! cc @bdhirsh @yanboliang

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 4]
```diff
@@ -233,6 +239,21 @@ def test_train_parity_multi_group(self):
...
+    @skip_if_lt_x_gpu(2)
+    @test_compiled_fsdp()
+    def test_train_parity_multi_group_compile(self):
+        self.run_subtests(
+            {
+                "reshard_after_forward": [True, False],
+                "device_type": ["cuda"],
+                "delay_after_forward": [False, True],
+                "delay_before_all_gather": [False],
+                "delay_before_reduce_scatter": [False],
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:251]: Are we only using `False` here to save test time?

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 5]
```diff
@@ -273,19 +294,25 @@ def _test_train_parity_multi_group(
         orig_reduce_scatter = dist.reduce_scatter_tensor
 
         def delayed_all_gather(*args, **kwargs):
-            if delay_before_all_gather:
-                torch.cuda._sleep(int(delay_in_ms * get_cycles_per_ms()))
+            torch.cuda._sleep(int(delay_in_ms * get_cycles_per_ms()))
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:297]: Curious why this change -- only stylistic or it affects compile?

File: torch/testing/_internal/common_fsdp.py [hunk 6]
```diff
@@ -1352,6 +1355,40 @@ def _test_fsdp_parity(
             )
 
 
+def test_compiled_fsdp(compile_compute_on_module: Optional[type] = None):
```
Comment (awgu) [inline torch/testing/_internal/common_fsdp.py:1358]: nit: Maybe we just `@no_type_check` this function since we are doing some patches, which mypy does not like and requires several `type: ignore` comments.

Comment (awgu) [inline torch/testing/_internal/common_fsdp.py:1358]: We may want to generalize this to a `Sequence[Type]` later (or some other container data structure).

**Review by awgu — APPROVED** [review]
Thanks!

File: torch/testing/_internal/common_fsdp.py (hunk 6 above)
Comment (weifengpy) [reply torch/testing/_internal/common_fsdp.py:1358]: you are right. will generalize this to ``Sequence`` in 2nd PR

Comment (weifengpy) [reply torch/testing/_internal/common_fsdp.py:1358]: nice! this should be cleaner. I improved patching to support ``fully_shard.state`` in an upcoming diff. Will see MYPY still complains

**Review by jansel — APPROVED** [review]

File: test/distributed/_composable/fsdp/test_fully_shard_training.py (hunk 4 above)
Comment (weifengpy) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:251]: there is a compiling issue for delayed all-gather/reduce-scatter. it's worth investigating (similar to reshard_after_forward=2)
``` File "/home/weif/local/pytorch/torch/distributed/_composable/fsdp/_fsdp_state.py", line 142, in _pre_forward
    args, kwargs = self._fsdp_param_group.pre_forward(module, args, kwargs)
  File "/home/weif/local/pytorch/torch/distributed/_composable/fsdp/_fsdp_param_group.py", line 212, in pre_forward
    self.wait_for_unshard()
  File "/home/weif/local/pytorch/torch/distributed/_composable/fsdp/_fsdp_param_group.py", line 175, in wait_for_unshard
    foreach_all_gather_copy_out(
  File "/home/weif/local/pytorch/torch/distributed/_composable/fsdp/_fsdp_collectives.py", line 80, in foreach_all_gather_copy_out
    all_gather_work.wait()
AttributeError: 'Tensor' object has no attribute 'wait'
```

File: test/distributed/_composable/fsdp/test_fully_shard_training.py (hunk 5 above)
Comment (weifengpy) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:297]: same to above. ``delayed_all_gather`` does not work in compiling. need to figure out the root cause
